File: 20170315-74.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """

        if not matrix:
            return False

        columns = len(matrix[0])
        rows = len(matrix)

        left = 0
        right = columns*rows - 1

        while left <= right:
            mid = (left+right)//2
            current = matrix[mid//columns][mid%columns]

            if current == target:
                return True
            elif current < target:
                left = mid + 1
            else:
                right = mid - 1

        return False

File: main/20170315-74.rb
# @param {Integer[][]} matrix
# @param {Integer} target
# @return {Boolean}
#
def search_matrix(matrix, target)
  return false if matrix.size == 0

  columns = matrix[0].size
  rows = matrix.size

  left = 0
  right = columns*rows - 1

  while left <= right
    mid = (left+right)/2
    current = matrix[mid/columns][mid%columns]

    case true
    when current == target
      return true
    when current < target
      left = mid + 1
    when current > target
      right = mid - 1
    end
  end

  return false
end
